draft/0005-impersonation.md

```diff
@@ -0,0 +1,333 @@
+---
+Start date: 2024-05-30
+Author(s): wangsijie
+Revision: 1
+---
+
+# Impersonation
+
+## Overview
+
+User impersonation is a feature that allows an authorized user to assume the identity of another user within the system. This capability can be useful for various reasons such as troubleshooting, providing support, and performing administrative tasks. This document aims to explore the design and implementation of user impersonation in Logto.
+
+## Motivation
+
+As an enterprise-ready identity provider, it is essential to provide user impersonation capabilities to support various use cases. For example, a support engineer may need to troubleshoot an issue on behalf of a user, or an administrator may need to verify the permissions of a user. User impersonation can be helpful to these processes and improve the overall user experience.
+
+## Token exchange
+
+The user impersonation feature will be implemented using the OAuth 2.0 [token exchange flow](https://www.rfc-editor.org/rfc/rfc8693.html). The token exchange flow allows an authorized user to exchange their access token for a new access token that represents the identity of another user. The new access token can then be used to make API requests on behalf of the impersonated user.
+
+### Request
+
+The client makes a token exchange request to the tenant's token endpoint with a special grant type using the HTTP POST method. The following parameters are included in the HTTP request entity-body using the `application/x-www-form-urlencoded` format.
+
+1. `grant_type`: REQUIRED. The value of this parameter must be `urn:ietf:params:oauth:grant-type:token-exchange` indicates that a token exchange is being performed.
+2. `resource`: OPTIONAL. The resource indicator, the same as other token requests.
+3. `scope`: OPTIONAL. The requested scopes, the same as other token requests.
+4. `subject_token`: REQUIRED. A security token that represents the identity of the party on behalf of whom the request is being made. See the next section for more details.
+5. `subject_token_type`: REQUIRED. The type of the security token provided in the `subject_token` parameter. The value of this parameter must be `urn:ietf:params:oauth:token-type:access_token`.
```

Comment:
Actually, I have an open question.  Since we are going to generate the subject_token by ourselves (not through OIDC-provider), what would be the exact definition for the type `urn:ietf:params:oauth:token-type:access_token`?
When can we state a token as an access_token and when cannot?

Reply:
i thought about this and figured it could be left to the developer's discretion. let the business decide.

it'll be great to specify this though.